### ISSUE TITLE
Improve the debugger when using prismatic/schema

### ISSUE THREAD

**Issue by expez** [clojure-emacs/cider#1642]
When you instrument a function which is annotated with type hints using [schema](https://github.com/plumatic/schema), using <kbd>C-u C-c C-c</kbd>, the debugger is immediately triggered and you have to hit <kbd>n</kbd> to step past all the type annotations.  After that, the function and the debugger interacts like normal.  If you hit <kbd>c</kbd> to skip past the type annotations, though, it's like the function never was instrumented.

In short, the debugger works as it should, you just have to endure a few extra keypresses.  It also took me a little while to realize it actually works, as I was hitting <kbd>c</kbd> when it asked me to step through the type annotations.

schema is very popular, and I personally find it invaluable as the system and/or team grows.

**Comment by Malabarba** [comment]
This happens with other special defs as well.
I don't think there's any general way to solve this, but adding a special case to `schema.core/def*` should be possible.

**Comment by Malabarba** [comment]
I've tuned the debugger to skip built-in schemas from schema.core, so at least you won't have to step through tirivial schemas.
Sadly, I can't think of a reasonable way to improve that further.

**Comment by expez** [comment]
@w01fe Any idea how can take this further?

**Comment by w01fe** [comment]
I'm happy to try to help, but I don't know anything about how the nrepl debugger works.  Can you explain a bit about how the debugger works and why it stops on the type annotations?

**Comment by Malabarba** [comment]
The cider debugger instruments the user code by wrapping sexps and symbols in a `breakpoint` macro.
This macro evaluates the content, shows it to the user, and waits for user input.

Thats basically how the debugger works. While code is executing, the debugger displays each new value to the user and pauses execution until the user types `n`. This is what we call "stepping through". Here's a screencast:
https://youtu.be/A3JAlWM8qRM

In order to be more helpful, there are some things that the debugger skips (as in, it proceeds with evaluation without displaying to the user). For instance, it skips values that are obviously constants (like a map literal). 

Ideally, we would also skip the type hints that are used with `schema.core/defn`. However, as far as I can tell, the debugger has no way of knowing these are type specifications. 
Like I said, the breakpoint macro is wrapped around the symbol that refers to a schema. So all the information it has access to is the symbol, its metadata, and the value it points to. So there's no way of knowing that this symbol/schema is being used as a type specification in a `s/defn`.

**Comment by bbatsov** [comment]
@Malabarba guess you can add some of this info to the manual for interested parties - e.g. in a section named `Internals` or something.

**Comment by w01fe** [comment]
Thanks for the context, I think I understand now -- so the debugger effectively works on the form before macroexpansion, and it's the mere presence of the schema forms (not the checking code they expand into) that are causing the problem?  Makes sense now that I think about it, since I guess stepping through a macroexpansion would not be especially illuminating.

In that case, it seems like doing this properly would require understanding the s/defn syntax and such.  Is there any explicit understanding of ordinary `defn` syntax (arglists, etc) already, or does the current behaviour just fall out of skipping the arglists and docstring automatically since they look like constants?  

Happy to do what I can to help, but it seems like it might be a pretty signifiant project (and not necessarily something that would even belong in cider itself?)?

**Comment by Malabarba** [comment]
Sort of... The debugger actually acts after macro expansion, but it does a lot of work before and after macroexpansion in order to be able to identify how the macroexpanded code relates to the original user-written code. This way it can step through the user-written code, but only has to worry about understanding macroexpanded code (which has no macros and only a dozen or so special forms). For instance, the debugger knows not to instrument the arglist of a `fn*`, and that automatically extends to the arglists of defn, defmacro, etc. 

The problem here is that schemas.core/defn extracts the schemas from the arglist into a let form, and so the debugger does see them. 

However, that's mostly irrelevant. I wouldn't want to add a ton of code to teach the debugger how to undertand a macroexpanded schemas.core/defn. I just wanted to know if schemas.core/defn adds some unique metadata to the arglist schemas. This would allow the debugger to easily spot these cases by just checking for this metadata.

**Comment by w01fe** [comment]
That's very helpful, thanks!  

Schema does not currently add any metadata to the arglist schemas, but we would be very happy to accept a patch that does so.  (I'm not currently set up to check that it works with the debugger).  

I think the three places to do it are:

https://github.com/plumatic/schema/blob/master/src/clj/schema/macros.clj#L142
https://github.com/plumatic/schema/blob/master/src/clj/schema/macros.clj#L152
https://github.com/plumatic/schema/blob/master/src/clj/schema/macros.clj#L253

(for ordinary args, rest args, and return schemas, respectively).

**Comment by Malabarba** [comment]
Thanks @w01fe!
I'll add this to my TODO list and we'll see how it goes.
